### PR TITLE
feat(server): efficient full app sync

### PR DESCRIPTION
**Pull request by fyfrey**
Adds two new endpoints (in a new controller  + service) to sync assets for the mobile app.

Compared to the old approach, these DB queries do not use offset/skip and can make use of the available indices. This should prevent the server/DB from overloading during initial/full app assets sync.

Further, the fast/delta sync now includes all owned+partner assets (thereby reducing the number of network requests and full table scans!)

This PR does not include the mobile changes. These will merged separately after 1-2 releases/weeks have passed to avoid all apps from breaking because the server has not yet been updated.

### CODE REVIEW

**Review by jrasm91 — COMMENTED** [review]
This looks pretty good. We're trying to avoid putting everything into the asset controller though. What do you think about splitting these new endpoints out into a controller and service specifically for syncing instead? Maybe endpoints like /sync/assets-full and /sync/assets-delta or similar?

File: server/src/dtos/asset.dto.ts [hunk 1]
```diff
@@ -130,3 +131,35 @@ export enum UploadFieldName {
...
+  @ValidateDate({ optional: true })
+  lastCreationDate?: Date;
+
+  @ValidateDate()
+  updatedUntil!: Date;
+
+  @IsInt()
+  @IsPositive()
+  @Type(() => Number)
+  limit!: number;
```
Comment (alextran1502) [inline server/src/dtos/asset.dto.ts:148]: We will need to use OpenApi decorator here as well for the property type

File: server/src/services/sync.service.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    if (!_.isEqual(userIds, dto.userIds) || duration > AUDIT_LOG_MAX_DURATION) {
+      // app does not have the correct partners synced
+      // or app has not synced in the last 100 days
+      return { needsFullSync: true, deleted: [], upserted: [] };
+    }
+
+    const limit = 10_000;
+    const upserted = await this.assetRepository.getChangedDeltaSync({ limit, updatedAfter: dto.updatedAfter, userIds });
+
+    if (upserted.length == limit) {
```
Comment (danieldietzler) [inline server/src/services/sync.service.ts:58]:
```suggestion
    if (upserted.length === limit) {
```

**Review by danieldietzler — APPROVED** [review]
LGTM

**Comment by alextran1502** [comment]
Hello @fyfrey, thank you for another high-quality PR. Can you add some words to the description about how the full sync vs delta sync works? What are the different mechanisms used between the two?

**Comment by fyfrey** [comment]
Sure, @alextran1502! It's actually the same as the current mechanism (just more performant on the server side):

Full sync: The app gets all asset information, compares this list with the state in it's database and upserts/removes entries as needed. The list of all assets is fetched in chunks per user, but still made into one giant list on the client (so, for 100k+ assets this is still an issue).

Delta sync: The app gets all changed and deleted assets for the user and partners. The server calculates this using the updatedAt property and the audit table (for deletes). So the app retrieves a comparable small set of changes regardless of the library size. It updates its database accordingly.

**Pull request merged**